scripts/build_test_history.py

```diff
@@ -0,0 +1,447 @@
+
+    try:
+        resp = requests.get(url)
+    # Sometimes this request can time out. If it does, just return
+    # an empty list and attempt the fetch again.
+    except requests.exceptions.ReadTimeout:
+        return []
+    runs_list: list[MetadataDict] = resp.json()
+
+    # If we have no runs to process in this date interval,
+    # we can skip this interval for processing from now on.
+    if len(runs_list) == 0:
+        print('No runs found for this interval.')
+        update_recent_processed_date(date_entity, end_interval_string)
+        # If we've processed up to (now - 24 hours), then return null,
+        # which signals we're done processing.
+        if end_interval == yesterday:
+            return None
+
+    # Sort by revision -> then time start, so that the aligned runs are
+    # processed in groups with each other.
+    # Note that this technically doesn't have an impact if only 1 set of
+    # aligned runs are processed, but this sort will allow this script to
+    # function properly if multiple aligned run sets were to be processed
+    # together.
+    runs_list.sort(key=lambda run: run['revision'])
+    runs_list.sort(key=lambda run: run['time_start'])
+
+    # Print the dates just to get info on the list of runs we're working with.
+    print('Runs to process:')
+    for run in runs_list:
+        print(f'ID: {run["id"]}, {run["browser_name"]} {run["time_start"]}')
+
+    return runs_list
+
+
+def update_recent_processed_date(
+        date_entity: MostRecentHistoryProcessed, new_date: str) -> None:
+    """Update the most recently processed date after finishing processing."""
+    date_entity.Date = new_date
+    date_entity.put()
+
+
+class NoRecentDateError(Exception):
+    """Exception raised when the MostRecentHistoryProcessed
+    entity is not found.
+    """
+    pass
+
+
+class JSONGenerationError(Exception):
+    """Exception raised when initial JSON files are being generated,
+    but the database has not been cleared of existing entries.
+    """
+    pass
+
+
+def get_processing_start_date() -> MostRecentHistoryProcessed:
+    most_recent_processed: MostRecentHistoryProcessed = (
+        MostRecentHistoryProcessed.query().get())
+
+    if most_recent_processed is None:
+        raise NoRecentDateError('Most recently processed run date not found.')
+    return most_recent_processed
+
+
+def check_if_db_empty() -> None:
+    """Raise an error if new JSON files are set to be generated and
+    test history data already exists.
+    """
+    test_history_entry: TestHistoryEntry = TestHistoryEntry.query().get()
+    if test_history_entry is not None:
+        raise JSONGenerationError(
+            'TestHistoryEntry entities exist in Datastore. '
+            'New JSON files should not be generated if data already exists.')
+
+
+# default parameters used for cloud functions.
+def main(args=None, topic=None) -> str:
+    client = ndb.Client(project=PROJECT_NAME)
+    with client.context():
```

Comment:
Can we add a verbose flag `-v` and add more debug logging? It will be easier to figure out problems through logging on cloud function, when something goes wrong

Reply:
Added!